### DISCUSSION TITLE
[QUESTION] SKBitmap.InstallPixels and general pixel memory ownership guidelines?

### DISCUSSION THREAD

**Discussion by ziriax** [mono/SkiaSharp#2044] [Q&A]
I'm having trouble to understand the ownership of pixel memory. 

For example,  the following code always saves a correct image A, and mostly saves a corrupt image B, or crashes randomly (at least in 1.68.3):

```cs
using System.IO;
using SkiaSharp;

namespace SkiaCropBug
{
	class Program
	{
		static SKBitmap Crop(int i)
		{
			using var codec = SKCodec.Create("kitten.jpg");
			using var bm1 = SKBitmap.Decode(codec);
			using var pm1 = bm1.PeekPixels();
			using var pm2 = pm1.ExtractSubset(SKRectI.Create(70, 20, 50, 45));

			var bm2 = new SKBitmap(pm2.Info, SKBitmapAllocFlags.None);
			bm2.InstallPixels(pm2);

			using var stream = File.Create($"frames/crop_{i:02}_A.png");
			bm2.Encode(stream, SKEncodedImageFormat.Png, 9);

			return bm2;
		}

		static void Main(string[] args)
		{
			Directory.CreateDirectory("frames");

			for (int i = 0; i < 100; ++i)
			{
				using var bm2 = Crop(i);
				using var stream = File.Create($"frames/crop_{i:02}_B.png");
				bm2.Encode(stream, SKEncodedImageFormat.Png, 9);
			}
		}
	}
}
```

Moving the `bm1` out of the loop fixes the problem, but now I would have to pass two bitmaps around, while surely it should be possible for `bm1` and `bm2` to share the same underlying `SkPixelRef`?

As a temporary workaround, I'm going to copy the pixels, but I would like to know the general rule here.

From the code above, it feels natural that ownership is managed for you, since I'm not passing any `IntPtr` around. However, looking the native code, `SKPixmap` doesn't store a `SKPixelRef`, only a pointer... 

So the managed bit is just an illusion, even though I'm coding C# here, I still have to think carefully about C/C++ ownership?

As far as I understand, the rules are:
- `SKPixmap` never frees memory, it is purely a (non reference counted!) view on the raw pixels, using a pointer
- `SKBitmap` and `SKImage` internally use an `SKPixelRef` (reference counted), but this reference counting is lost as soon as a pixmap is involved...
- So basically, one should never create a bitmap or image from a pixmap, unless this creation copies the pixels over.

Is this correct?

**Comment by ziriax** [comment]
Mmm, it seems in this case, I could have used the `SKBitmap.ExtractSubSet` method, that does share the underlying native `SkPixelRef`...
